File: 2015/cpp/day15/part1.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <algorithm>
#include <vector>
#include <stdlib.h>

using namespace std;

vector<string> getLines() {
    ifstream file("input.txt");
    vector<string> lines;
    string line;

    while (getline(file, line))
        lines.push_back(line);

    return lines;
}

void istringstreamExample(string line) {

}

int main () {
    vector<string> lines = getLines();
    int ingredients[4][4];

    for(int i = 0; i < lines.size(); i++) {
        replace(lines.at(i).begin(), lines.at(i).end(), ',', ' ');

        istringstream iss(lines.at(i));
        string garbage;

        iss >> garbage >> garbage >> ingredients[i][0] >> garbage >> ingredients[i][1]
            >> garbage >> ingredients[i][2] >> garbage >> ingredients[i][3];
    }

    int record = 0;

    for (int sugar = 0; sugar < 100; sugar++) {
        for (int sprinkles = 0; sprinkles < 100 - sugar; sprinkles++) {
            for (int candy = 0; candy < 100 - sugar - sprinkles; candy++) {
                int chocolate = 100 - sugar - sprinkles - candy;

                int total = 1;

                for (int ingredient = 0; ingredient < 4; ingredient++) {
                    int subTotal = ingredients[0][ingredient] * sugar +
                                   ingredients[1][ingredient] * sprinkles +
                                   ingredients[2][ingredient] * candy +
                                   ingredients[3][ingredient] * chocolate;

                    if (subTotal <= 0) {
                        total = 0;
                        break;
                    }

                    total *= subTotal;
                }

                record = max(record, total);
            }
        }
    }

    cout << record << endl;
}

File: 2015/ruby/day15/part1.rb
#!/usr/bin/ruby

begin
    File.open('input.txt', 'r') do |file|
        ingredients = {}
        max = 0

        file.each_line do |line|
            stats = line.delete(':,').downcase.split(' ')

            ingredients[stats[0]] = {
                :capacity => stats[2].to_i,
                :durability => stats[4].to_i,
                :flavor => stats[6].to_i,
                :texture => stats[8].to_i
            }
        end

        for sugar in 0..100
            for sprinkles in 0..(100 - sugar)
                for candy in 0..(100 - sugar - sprinkles)
                    chocolate = 100 - sugar - sprinkles - candy

                    capacity = (
                        sugar * ingredients['sugar'][:capacity] +
                        sprinkles * ingredients['sprinkles'][:capacity] +
                        candy * ingredients['candy'][:capacity] +
                        chocolate * ingredients['chocolate'][:capacity]
                    )
                    durability = (
                        sugar * ingredients['sugar'][:durability] +
                        sprinkles * ingredients['sprinkles'][:durability] +
                        candy * ingredients['candy'][:durability] +
                        chocolate * ingredients['chocolate'][:durability]
                    )
                    flavor = (
                        sugar * ingredients['sugar'][:flavor] +
                        sprinkles * ingredients['sprinkles'][:flavor] +
                        candy * ingredients['candy'][:flavor] +
                        chocolate * ingredients['chocolate'][:flavor]
                    )
                    texture = (
                        sugar * ingredients['sugar'][:texture] +
                        sprinkles * ingredients['sprinkles'][:texture] +
                        candy * ingredients['candy'][:texture] +
                        chocolate * ingredients['chocolate'][:texture]
                    )

                    capacity = capacity < 0 ? 0 : capacity
                    durability = durability < 0 ? 0 : durability
                    flavor = flavor < 0 ? 0 : flavor
                    texture = texture < 0 ? 0 : texture

                    total = capacity * durability * flavor * texture
                    max = total > max ? total : max
                end
            end
        end

        puts max
    end
rescue SystemCallError
    puts 'input.txt not found.'
end
